File: kotlin/shells-compose/src/main/kotlin/dev/modaal/duet/shells/Working.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.duet.shells

import kotlinx.coroutines.awaitCancellation

// The worker seam: lifecycle-bound stateful processing, re-substrated on
// structured concurrency (doc 20). A worker is an environment-side object that
// may own mutable state and long-lived subscriptions, bracketed by a host's
// mount lifetime, feeding features ONLY through the existing seams —
// environment Flows consumed by effect handlers, or actions through a
// shell-bound Relay. Workers never touch stores or feature state ("no silent
// ingress"); workers process, reducers decide (R16).
//
// Config-change stance (Q2): Android workers live in the RETAINED/logical
// component scope (the tree that survives rotation — InstanceKeeper-backed
// under Decompose), so "one bracket per mount" means LOGICAL mount on both
// platforms. View-bound scopes stay for view-side observation only.

/**
 * One worker: `run()` is its whole life. A host starts it at adoption
 * ([StoreHost.adopt]) and cancellation of the launched coroutine IS stop —
 * there is no separate `stop()` to forget. One bracket per (logical) mount: a
 * remount adopts a fresh instance; "pause when backgrounded" is consumed as an
 * input stream, never a lifecycle callback.
 *
 * Implementation guidance, not API: stateful workers confine their state to
 * the host's (main-confined) scope — isolation replaces hand-rolled locks. The
 * `run()` body and its structured children are a sanctioned concurrency home;
 * a worker launching unstructured jobs is the same review defect it would be
 * in feature code. Swift-flavor mirror: DuetShells/Working.swift.
 */
fun interface Working {
  /**
   * The worker's whole life, from adoption to host teardown. MUST return
   * promptly once cancelled — a worker still running after cancellation is a
   * leak (`WorkerTester.finish()` fails on it, and the host's
   * [StoreHost.liveWorkerCount] ledger stays nonzero).
   */
  suspend fun run()
}

/**
 * The bracket tail for subscription-shaped workers: set the subscriptions up in
 * `run()`, then `untilCancelled()`. Thinness note: on Kotlin this is
 * [awaitCancellation] — structured concurrency provides the park; the Swift
 * twin needs a hand-rolled continuation.
 */
suspend fun untilCancelled(): Nothing = awaitCancellation()

File: swift/Sources/DuetShells/Working.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import Foundation

// The worker seam: lifecycle-bound stateful processing, re-substrated on
// structured concurrency. A worker is an environment-side object that may own
// mutable state and long-lived subscriptions, bracketed by a host's mount
// lifetime, feeding features ONLY through the existing seams — environment
// streams consumed by effect handlers, or actions through a shell-bound Relay.
// Workers never touch stores or feature state ("no silent ingress"), and a
// worker that grows decision logic is asking to move it into a reducer:
// workers process, reducers decide.
//
// Deliberately freestanding: no kernel type appears in this file, so a legacy
// app can adopt the worker bracket before it adopts the kernel.

/// One worker: `run()` is its whole life. A host starts it at adoption
/// (`StoreHost.adopt`) and cancellation of the surrounding task IS stop —
/// there is no separate `stop()` to forget. One bracket per mount: a remount
/// adopts a fresh instance; "pause when backgrounded" is consumed as an input
/// stream (scenePhase is data), never a lifecycle callback.
///
/// Implementation guidance, not API: stateful workers are `actor`s, or
/// `@MainActor` classes when main-bound — isolation replaces hand-rolled
/// serial queues and locks. The `run()` body and its structured children are
/// a sanctioned concurrency home; a worker spawning unstructured `Task {}`s
/// is the same review defect it would be in feature code.
public protocol Working: AnyObject, Sendable {
  /// The worker's whole life, from adoption to host teardown. MUST return
  /// promptly once the surrounding task is cancelled — a worker still running
  /// after cancellation is a leak (`WorkerTester.finish()` fails on it, and
  /// the host's `liveWorkerCount` ledger stays nonzero).
  func run() async
}

extension Working {
  /// The bracket tail for workers whose machinery is subscription-shaped
  /// (Combine pipelines, callback registries) rather than loop-shaped: set up
  /// the subscriptions in `run()`, then `await untilCancelled()`. Returns when
  /// the surrounding task is cancelled; locals then release, cancelling the
  /// subscriptions — teardown stays structured without a `stop()` hook.
  public func untilCancelled() async {
    let park = Park()
    await withTaskCancellationHandler {
      await park.wait()
    } onCancel: {
      park.cancel()
    }
  }
}

/// The parked continuation behind `untilCancelled()`, tolerant of the
/// cancel-before-wait race. Lock discipline: the lock is only ever taken in
/// synchronous scopes (the continuation body and `cancel()`), never across a
/// suspension.
private final class Park: @unchecked Sendable {
  private let lock = NSLock()
  private var continuation: CheckedContinuation<Void, Never>?
  private var cancelled = false

  func wait() async {
    await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
      lock.lock()
      if cancelled {
        lock.unlock()
        cont.resume()
      } else {
        continuation = cont
        lock.unlock()
      }
    }
  }

  func cancel() {
    lock.lock()
    cancelled = true
    let cont = continuation
    continuation = nil
    lock.unlock()
    cont?.resume()
  }
}
